### PR TITLE
Add input of country and city through the URL

### PR DESCRIPTION
**Pull request by artem-str-31**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -33,20 +33,28 @@ def index():
...
-        last_result = api_fetch()
+        last_result = api_fetch("UK", "London")
         last_fetch_delta = 0
     return f"Temperature: {last_result.temp} Time since updated: {last_fetch_delta}"
 
-def api_fetch() -> WeatherData:
+
+@app.route('/<country_name>/<city_name>')
+def city_weather(country_name, city_name):
+    result = api_fetch(country_name, city_name)
```
Comment (will-lynas) [inline app.py:46]: Now that we are using user supplied data in the api call, we need to handle errors from the api. Make `api_fetch` throw an exception if the api doesn't support a city. Catch the exception in `city_weather` and display a suitable response.

File: app.py [hunk 2]
```diff
@@ -33,20 +33,28 @@ def index():
...
+
+
+def api_fetch(country_name, city_name) -> WeatherData:
     with open('api_key.txt') as f:
         api_key = f.read().strip()
     params = {
-            "q": "London,uk",
-            "APPID": api_key
-            }
+            "q": f"{city_name}, {country_name}",
```
Comment (will-lynas) [inline app.py:54]: Is the space needed? Does it care if there is a space or not?

**Review by will-lynas — CHANGES_REQUESTED** [review]

File: app.py [hunk 3]
```diff
@@ -43,8 +45,19 @@ def index():
 
 @app.route('/<country_name>/<city_name>')
 def city_weather(country_name, city_name):
-    result = api_fetch(country_name, city_name)
-    return f"Country: {country_name}. City: {city_name}. Temperature: {result.temp}."
+    last_result = db.session.query(WeatherData).filter(WeatherData.country.like(country_name), WeatherData.city.like(city_name)).order_by(WeatherData.time.desc()).first()
```
Comment (will-lynas) [inline app.py:31]: I don't think we need `like` here since we want an exact match. `==` instead.

File: app.py [hunk 4]
```diff
@@ -24,27 +26,21 @@ def __repr__(self):
 with app.app_context():
     db.create_all()
 
-@app.route('/')
-def index():
-    last_result = db.session.query(WeatherData).order_by(WeatherData.time.desc()).first()
+@app.route('/<country_name>/<city_name>')
+def city_weather(country_name, city_name):
+    last_result = db.session.query(WeatherData).filter(WeatherData.country.like(country_name), WeatherData.city.like(city_name)).order_by(WeatherData.time.desc()).first()
```
Comment (will-lynas) [inline app.py:31]: I think it would be better to have the caching stuff in `api_fetch` or in a separate function so that it can be reused.

Comment (will-lynas) [inline app.py:31]: Line too long. Add newlines somewhere. We will enforce this with `pylint` in future.

File: app.py [hunk 5]
```diff
@@ -57,7 +53,7 @@ def api_fetch(country_name, city_name) -> WeatherData:
         }
     endpoint = "http://api.openweathermap.org/data/2.5/weather"
     response = requests.get(endpoint, params=params).json()
-    weather_data = WeatherData(temp=response["main"]["temp"], time=time.time())
+    weather_data = WeatherData(country=country_name, city=city_name, temp=response["main"]["temp"], time=time.time())
```
Comment (will-lynas) [inline app.py:56]: Newlines

File: app.py [hunk 6]
```diff
@@ -24,27 +26,21 @@ def __repr__(self):
 with app.app_context():
     db.create_all()
 
-@app.route('/')
-def index():
```
Comment (will-lynas) [inline app.py:28]: Make index redirect to `/UK/London`

**Review by will-lynas — CHANGES_REQUESTED** [review]

File: app.py [hunk 7]
```diff
@@ -23,27 +27,35 @@ class WeatherData(db.Model):
     def __repr__(self):
         return f"<{self.__class__.__name__} {self.temp} {self.time}>"
```
Comment (will-lynas) [inline app.py:28]: Can you add `city` and `country` to the `__repr__`

File: app.py [hunk 8]
```diff
@@ -23,27 +27,35 @@ class WeatherData(db.Model):
...
+def city_weather(country_name, city_name): 
+    try:
+        result = api_fetch(country_name, city_name)
+    except CityNotFoundError:
+        return 'City not Found', 404
+    return f"Country: {country_name}. City: {city_name}. Temperature: {result.temp}. Time since updated: {time.time() - result.time}"
+
+
+def api_fetch(country_name, city_name) -> WeatherData:
+    last_result = db.session.query(WeatherData).filter(WeatherData.country.is_(country_name), WeatherData.city.is_(city_name)).order_by(WeatherData.time.desc()).first()
```
Comment (will-lynas) [inline app.py:51]: afaik `is_` is the equivalent of sql `IS` which is for checking for null. I think you literally want `==`.

File: app.py [hunk 9]
```diff
@@ -52,8 +64,12 @@ def api_fetch(country_name, city_name) -> WeatherData:
...
-    response = requests.get(endpoint, params=params).json()
-    weather_data = WeatherData(country=country_name, city=city_name, temp=response["main"]["temp"], time=time.time())
+    response = requests.get(endpoint, params=params)
+    if not response:
+        raise CityNotFoundError
+    response_json = response.json()
+    weather_data = WeatherData(country=country_name, city=city_name, temp=response_json["main"]["temp"], time=time.time())
     db.session.add(weather_data)
     db.session.commit()
+    last_fetch_delta = 0
```
Comment (will-lynas) [inline app.py:74]: This line doesn't do anything.

**Review by will-lynas — CHANGES_REQUESTED** [review]

**Review by will-lynas — APPROVED** [review]
Merge (as long as it works locally)

**Comment by will-lynas** [comment]
And make sure to squash first. `rebase -i origin/main` then `push -f`.

**Comment by artem-str-31** [comment]
> And make sure to squash first. `rebase -i origin/main` then `push -f`.
> Or use github.

Should I squash and merge or rebase and merge?

**Comment by will-lynas** [comment]
`rebase` doesn't actually do anything here. Generally you want to do both, so better to use cmdline. Here you can just you squash and merge.

**Pull request merged**